scripts/bundle-images.sh

```diff
@@ -0,0 +1,245 @@
+        PROVIDER_KIND_TOLOWER="control-plane"
+    fi
+
+    # coreprovider does not have a provider prefix.
+    if [[ $PROVIDER_KIND_TOLOWER == "coreprovider" ]]; then
+        LABEL_VALUE=$(echo $PROVIDER_NAME)
+    else
+        LABEL_VALUE=$(echo $(echo $PROVIDER_KIND_TOLOWER | sed -e 's/provider//g')-$PROVIDER_NAME)
+    fi
+
+    wait_for_deploy_exist "$LABEL_KEY=$LABEL_VALUE"
+
+    $KUBECTL wait --for condition=available --timeout=2m deploy -l $LABEL_KEY=$LABEL_VALUE -n $NAMESPACE
+    if [[ $? -ne 0 ]]; then
+        echo "Error: Cannot wait for Deployment: Deployment with $LABEL_KEY=$LABEL_VALUE label not found"
+        exit 1
+    fi
+done
+
+
+# Now that we know everything is deployed and ready, we can get all of images by
+# execing into the KIND cluster.
+CONTROL_PLANE=$($KUBECTL get nodes --no-headers -o custom-columns=":metadata.name")
+if [[ $? -ne 0 ]] || [[ $CONTROL_PLANE == "" ]]; then
+    echo "Error: Cannot get control plane node"
+    exit 1
+fi
+
+echo -e "\nPulling images for HMC components...\n"
+
+for IMAGE in $(docker exec -it $CONTROL_PLANE crictl images | sed 1,1d | awk '{print $1":"$2}' | grep -v 'kindest');
+do
+    if [[ $IMAGE == "" ]]; then
+        echo "Error: Failed to get image from KIND cluster, image string should not be empty"
+        exit 1
+    fi
+
+    if [[ $IMAGE == "docker.io/hmc/controller:latest" ]]; then
+        # Don't try to pull and package the local controller image.
+        continue
+    fi
+
+    docker pull $IMAGE
+    if [[ $? -ne 0 ]]; then
+        echo "Error: Failed to pull $IMAGE"
+        exit 1
+    fi
+
+    IMAGES_BUNDLED="$IMAGES_BUNDLED $IMAGE"
+done
+
+echo -e "\nPulling images for HMC extensions...\n"
+
+# Next, we need to build a list of images used by k0s extensions.  Walk the
+# templates directory and extract the images used by the extensions.
+for TEMPLATE in $(find $TEMPLATES_DIR -name 'k0s*.yaml');
+do
+    if [[ $TEMPLATE == *"k0smotron"* ]]; then
+        EXTENSIONS_PATH=".spec.k0sConfig.spec.extensions.helm"
+    else
+        EXTENSIONS_PATH=".spec.k0sConfigSpec.k0s.spec.extensions.helm"
+    fi
+
+    REPOS=$(grep -vw "{{" $TEMPLATE | $YQ e "${EXTENSIONS_PATH}.repositories[] | [.url, .name] | join(\";\")")
+    for REPO in $REPOS
+    do
+        URL=$(echo $REPO | cut -d';' -f1)
+        NAME=$(echo $REPO | cut -d';' -f2)
+
+        VERSION=$(grep -vw "{{" $TEMPLATE |
+            $YQ e "${EXTENSIONS_PATH}.charts[] | select(.name == \"$NAME\") | .version")
+
+        VALUES=$(grep -vw "{{" $TEMPLATE |
+            $YQ e "${EXTENSIONS_PATH}.charts[] | select(.name == \"$NAME\") | .values")
+
+        if [[ $URL == "" ]] || [[ $NAME == "" ]] || [[ $VERSION == "" ]]; then
+            echo "Error: Failed to get URL, NAME, or VERSION from $TEMPLATE"
+            exit 1
+        fi
+
+        # Check to see if a custom image is being used.
+        CUSTOM_TAG=$(echo -e "$VALUES" | $YQ e '.image.tag' -)
```

Comment:
Can we just pass values to `helm template` and omit all `custom_tag` logic?

Reply:
Yep, good idea, I've implemented this.